README.md
# buffer-rpc

[![Build Status](https://travis-ci.org/bufferapp/buffer-rpc.svg?branch=master)](https://travis-ci.org/bufferapp/buffer-rpc)

Buffer RPC request handler

## Quickstart

Create a RPC method to add 2 numbers:

```js
// index.js
const express = require('express')
const { rpc, method } = require('@bufferapp/buffer-rpc')
const app = express()
app.use(bodyParser.json()) // this is required
app.post('/rpc/:method?', rpc(method('add', (a, b) => a + b)))
const port = 3000
app.listen(port, () => console.log(`App Is Listening On Port ${port}`))
```

Start the server

```sh
node index.js
```

Or you can use curl to call the `add` method:

```sh
curl -H "Content-Type: application/json" -X POST -d '{"args": "[2, 3]"}' localhost:3000/add | python -m json.tool

# {
#    "result": 5
# }
```

_Note: it is recommended that you use the RPC client:_

https://github.com/bufferapp/micro-rpc-client

To see a list of all available methods use the `methods` call:

```sh
curl -H "Content-Type: application/json" -X POST -d '{"name": "methods"}' localhost:3000 | python -m json.tool

# {
#   result: [
#     {
#       "docs": "add two numbers"
#       "name": "add"
#     },
#     {
#       "docs": "list all available methods",
#       "name": "methods"
#     }
#   ]
# }
```

## Usage

Here's a few examples of how to hook up the handler methods:

```js
const express = require('express')
const { rpc, method } = require('@bufferapp/buffer-rpc')
const app = express()
app.use(bodyParser.json()) // this is required

app.post(
  '/rpc/:method?',
  rpc(
    method('add', (a, b) => a + b),
    method(
      'addAsync',
      (a, b) =>
        new Promise(resolve => {
          resolve(a + b)
        }),
    ),
    method('addItems', ({ a, b }) => a + b),
    method(
      'addItemsAsync',
      ({ a, b }) =>
        new Promise(resolve => {
          resolve(a + b)
        }),
    ),
    method('throwError', () => {
      throw createError({ message: "I'm sorry I can't do that" })
    }),

    method(
      'throwErrorAsync',
      () =>
        new Promise((resolve, reject) => {
          reject(
            createError({
              message: 'Something is broke internally',
              statusCode: 500,
            }),
          )
        }),
    ),
    method(
      'documentation',
      `
  # documentation

  Document what a method does.
  `,
      () => 'documentation',
    ),
  ),
)

const port = 3000
app.listen(port, () => console.log(`App Is Listening On Port ${port}`))
```

## Dependency Injection

To simplify the code in your RPC methods, you may want to pass some common utilities down via simple dependancy injection. To do this, you call the `rpc()` method with slightly different syntax.

Instead of this (seen in the other README examples):

```js
rpc(methodOne, methodTwo, methodThree, ...);
```

Pass the methods as an `Array` and pass `utils` as a second parameter:

```js
rpc([methodOne, methodTwo, methodThree], utils);
```

Where `utils` is an `Object` that will be exposed to your RPC methods as the last parameter. For example, you might use it like this (example is simplified):

```js
// rpcHandler.js
const { rpc } = require('@bufferapp/buffer-rpc');

const PublishAPI = require('./publishAPI');
const myMethod = require('./myMethod');

module.exports = rpc([myMethod], { PublishAPI });

// myMethod.js
const { method } = require('@bufferapp/buffer-rpc');

module.exports = method(
  'myMethod',
  'myMethodDocs',
  (args, req, res, { PublishAPI }) => PublishAPI.fetch('/1/user.json')
);
```

## Error Handling

### Handled Flag

To help understand what the handled flag is for lets talk about a simplified distributed system. 

```
client -> server
```

The server is the RPC endpoints and the client is the source of the request, perhaps a browser or another service.. When handled = true this means there is nothing more for the client to do, when handled = false this means there _might_ be more work to resolve an issue. So handled = false is a maybe. An example of a definite handled = true is when you make a request to an RPC endpoint that does not exist. There's no partial state to resolve. An example of handled = false would be when you have to make multiple writes to a database, the first one passes and the second one fails. You've got a partially handled failure case. The client would be notified with a handled = false and then resolve the issue by either deleting the first record or writing the second -- depends on what the application does!

While this doesn't happen very often, it is more likely when we choose to use a NoSQL database since we don't have joins. That being said most of the time, you'll return an error with handled = true... especially if you're keeping RPC endpoints simple. Avoid distributed systems problems when you can!

**createError** handled = true (customizable)

**errorMiddleware** handled = false

**method not found** handled = true

### Error Codes

These are the default error codes for error type responses

**1000** - createError default (customizable)  
**4040** - method not found (404)  
**5000** - unhandled exception (500)

## API

### rpc

Takes a bunch of methods as arguments. Passes requests to right RPC endpoint

```js
rpc(...methods)
```

**...methods** - _method_ - rpc method (see below)

### method

add a remote method

```js
method(name, [docs], fn)
```

**name** - _string_ - the name of the method  
**docs** - _string_ - documentation about a method  
**fn** - _function_ - the function to call and apply parameters the method is requested

### createError

create an error to be thrown, optionally set the status code

```js
createError({
  message,
  code = 1000,
  statusCode = 400,
  handled = true,
})
```

**message** - _string_ - error message to return  
**code** - _integer_ - custom error code to add to response body HTTP status code (default to 1000)  
**statusCode** - _integer_ - HTTP status code (default to 400)
**handled** - _boolean_ - add if error was handled on backend to the response body (default to true)

### errorMiddleware

[express/connect error handling middleware](https://expressjs.com/en/guide/error-handling.html) that receives and unhandled error and returns a JSON response

```js
app.post('/rpc', rpc(method('awake', () => 'OK')), errorHandler)

/*
statusCode = 500
body = {
  handled: false,
  code: 5000,
  error: 'some error message' // set from error.message
}
*/
```

## Request and Response Objects

Request and response objects are always passed along as the last two arguments in case they're needed.

```js
method('addWithSession', (a, b, req, res) => {
  if (!req.session) {
    throw createError({ message: 'a session is needed to add numbers', statusCode: 401})
  }
  return a + b
}
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.1.0]
### Added
- When composing your RPC handler you can optionally pass an dependencies to be exposed to all RPC methods. See the README section on "Dependency Injection" for more details. This support is optional and existing usage will continue to function.

## [1.0.0]
### Added
- `buffer-rpc` can now receive calls in the form of `endpoint/:methodName` to allow for clearer debugging and tracing (e.g., in Chrome DevTools and DataDog). The previous method of passing the name to the RPC endpoint as one of the JSON body parameters **is still supported** but the method name in the URI takes precedence. 

To upgrade your application do the following:

1. In your server, upgrade `@bufferapp/buffer-rpc` to version `1.0.0`.
2. If your front-end app uses `@bufferapp/async-data-fetch` then upgrade that to version `2.0.0`.
3. If your front-end uses the RPC Client directly, then upgrade `@bufferapp/micro-rpc-client` to version `1.0.0`.
4. In your express server, change the route handler for your RPCs to support the new `:method` wildcard. For example, from this:

```
// ❌
app.post('/rpc', checkToken, rpcHandler, errorMiddleware);
```

  To this:

```
// ✅
app.post('/rpc/:method?', checkToken, rpcHandler, errorMiddleware);
```

5.  If you have a `checkToken` method as in the one above, you'll also need to update that. From this:

```
// ...
module.exports = (req, res, next) => {
  const { name } = req.body;
```

  To this:

```
// ...
module.exports = (req, res, next) => {
  const name = req.params.method ? req.params.method : req.body.name;
```

## [Unreleased]
### Added
- Integration tests for [micro-rpc-client](https://github.com/bufferapp/micro-rpc-client) error handling

### Changed
- README update

## [0.0.4] - 2018-10-10
### Added
- Error handler middleware

## [0.0.3] - 2018-10-04
### Added
- Optional custom error code

## [0.0.2] - 2018-09-11

- Initial public release


test/utils.js
const http = require('http')
const express = require('express')
const bodyParser = require('body-parser')
const request = require('request-promise')

const createServer = (handler, errorHandler) => {
  const app = express()
  app.use(bodyParser.json())
  if (errorHandler) {
    app.post('/rpc/:method?', handler, errorHandler)
  } else {
    app.post('/rpc/:method?', handler)
  }
  return http.createServer(app)
}

const stopServer = server => server.close()

const generateRequest = ({ url, name, args }) => {
  return request({
    uri: `${url}/rpc/${name}`,
    method: 'POST',
    body: {
      args: JSON.stringify(args),
    },
    json: true,
  })
}

module.exports = {
  createServer,
  stopServer,
  generateRequest,
}


test/rpc.test.js
const http = require('http')
const listen = require('test-listen')
const express = require('express')
const request = require('request-promise')
const { createServer, stopServer, generateRequest } = require('./utils')
const createError = require('../src/createError')
const rpc = require('../src/rpc')

describe('rpc', () => {
  it('should handle unknown method', async () => {
    expect.assertions(2)
    const server = createServer(rpc())
    let url = await listen(server)
    try {
      await generateRequest({
        url,
        name: 'unknown',
      })
    } catch (error) {
      expect(error.statusCode).toBe(404)
      expect(error.error).toEqual({
        error: 'unknown method',
        handled: true,
        code: 4040,
      })
    }
    stopServer(server)
  })
  it('should handle a request to methods', async () => {
    const server = createServer(rpc())
    let url = await listen(server)
    const body = await generateRequest({
      url,
      name: 'methods',
    })
    expect(body).toEqual([
      {
        name: 'methods',
        docs: 'list all available methods',
      },
    ])
    stopServer(server)
  })
  it('should handle a request to methods with custom methods', async () => {
    const name = 'name'
    const fn = () => 'OK'
    const docs = 'some custom method'
    const method = {
      name,
      docs,
      fn,
    }
    const server = createServer(rpc(method))
    let url = await listen(server)
    const body = await generateRequest({
      url,
      name: 'methods',
    })
    expect(body).toEqual([
      {
        name: 'methods',
        docs: 'list all available methods',
      },
      {
        name,
        docs,
      },
    ])
    stopServer(server)
  })
  it('should handle a request to a new method', async () => {
    const name = 'name'
    const result = 'hello, world'
    const fn = () => result
    const method = {
      name,
      fn,
    }
    const server = createServer(rpc(method))
    let url = await listen(server)
    const body = await generateRequest({
      url,
      name,
    })

    expect(body).toEqual({ result })
    stopServer(server)
  })
  it('should handle request to a method with an async function', async () => {
    const name = 'name'
    const result = 'hello, world - async'
    const fn = async () => {
      return result
    }
    const method = {
      name,
      fn,
    }
    const server = createServer(rpc(method))
    let url = await listen(server)
    const body = await generateRequest({
      url,
      name,
    })
    expect(body).toEqual({ result })
    stopServer(server)
  })

  it('should handle request to a method with an unhandled failing async function', async () => {
    expect.assertions(2)
    const name = 'name'
    const errorMessage = 'nope'
    const fn = async () => {
      throw new Error(errorMessage)
    }
    const method = {
      name,
      fn,
    }
    const errorHandler = (error, req, res, next) => {
      if (res.headersSent) {
        return next(error)
      }
      res.status(500).send({ error: error.message })
    }
    const server = createServer(rpc(method), errorHandler)
    try {
      let url = await listen(server)
      await generateRequest({
        url,
        name,
      })
    } catch (error) {
      expect(error.statusCode).toBe(500)
      expect(error.error).toEqual({
        error: errorMessage,
      })
    }
    stopServer(server)
  })

  it('should handle request to a method with an unhandled failing function', async () => {
    expect.assertions(2)
    const name = 'name'
    const errorMessage = 'nope'
    const fn = () => {
      throw new Error(errorMessage)
    }
    const method = {
      name,
      fn,
    }
    const errorHandler = (error, req, res, next) => {
      if (res.headersSent) {
        return next(error)
      }
      res.status(500).send({ error: error.message })
    }
    const server = createServer(rpc(method), errorHandler)
    try {
      let url = await listen(server)
      await generateRequest({
        url,
        name,
      })
    } catch (error) {
      expect(error.statusCode).toBe(500)
      expect(error.error).toEqual({
        error: errorMessage,
      })
    }
    stopServer(server)
  })

  it('should handle request to a method with a handled failing function', async () => {
    expect.assertions(2)
    const name = 'name'
    const errorMessage = 'nope'
    const fn = () => {
      throw createError({
        message: errorMessage,
      })
    }
    const method = {
      name,
      fn,
    }
    const server = createServer(rpc(method))
    try {
      let url = await listen(server)
      await generateRequest({
        url,
        name,
      })
    } catch (error) {
      expect(error.statusCode).toBe(400)
      expect(error.error).toEqual({
        error: errorMessage,
        code: 1000,
        handled: true,
      })
    }
    stopServer(server)
  })

  it('should handle request to a method with a handled async failing function', async () => {
    expect.assertions(2)
    const name = 'name'
    const errorMessage = 'nope'
    const fn = async () => {
      throw createError({
        message: errorMessage,
      })
    }
    const method = {
      name,
      fn,
    }
    const server = createServer(rpc(method))
    try {
      let url = await listen(server)
      await generateRequest({
        url,
        name,
      })
    } catch (error) {
      expect(error.statusCode).toBe(400)
      expect(error.error).toEqual({
        error: errorMessage,
        code: 1000,
        handled: true,
      })
    }
    stopServer(server)
  })

  it('should handle request to a method with custom status code', async () => {
    expect.assertions(2)
    const name = 'name'
    const errorMessage = 'nope'
    const statusCode = 401
    const fn = async () => {
      throw createError({
        message: errorMessage,
        statusCode,
      })
    }
    const method = {
      name,
      fn,
    }
    const server = createServer(rpc(method))
    try {
      let url = await listen(server)
      await generateRequest({
        url,
        name,
      })
    } catch (error) {
      expect(error.statusCode).toBe(statusCode)
      expect(error.error).toEqual({
        error: errorMessage,
        code: 1000,
        handled: true,
      })
    }
    stopServer(server)
  })

  it('should handle a request with name in header', async () => {
    const name = 'name'
    const result = 'hello, world'
    const fn = () => result
    const method = {
      name,
      fn,
    }
    const server = createServer(rpc(method))
    let url = await listen(server)
    const body = await request({
      uri: `${url}/rpc`,
      method: 'POST',
      headers: {
        'x-buffer-rpc-name': name,
      },
      json: true,
    })

    expect(body).toEqual({ result })
    stopServer(server)
  })

  it('should handle a request with name in the body', async () => {
    const name = 'name'
    const result = 'hello, world'
    const fn = () => result
    const method = {
      name,
      fn,
    }
    const server = createServer(rpc(method))
    let url = await listen(server)
    const body = await request({
      uri: `${url}/rpc`,
      method: 'POST',
      body: {
        name: 'name',
      },
      json: true,
    })

    expect(body).toEqual({ result })
    stopServer(server)
  })

  it('should handle a request with arguments', async () => {
    const name = 'name'
    const fn = (a, b) => a + b
    const method = {
      name,
      fn,
    }
    const server = createServer(rpc(method))
    let url = await listen(server)
    const body = await generateRequest({
      url,
      name,
      args: [2, 2],
    })

    expect(body).toEqual({ result: 4 })
    stopServer(server)
  })

  it('should handle a request with keyed arguments', async () => {
    const name = 'name'
    const fn = ({ a, b }) => a + b
    const method = {
      name,
      fn,
    }
    const server = createServer(rpc(method))
    let url = await listen(server)
    const body = await generateRequest({
      url,
      name,
      args: { a: 3, b: 1 },
    })

    expect(body).toEqual({ result: 4 })
    stopServer(server)
  })

  it('should fail if parsed body is missing', async () => {
    expect.assertions(2)
    const app = express()
    app.post('*', rpc())
    app.use((error, req, res, next) => {
      if (res.headersSent) {
        return next(error)
      }
      res.status(500).send({ error: error.message })
    })
    const server = http.createServer(app)
    let url = await listen(server)
    try {
      await generateRequest({
        url,
        name: 'methods',
      })
    } catch (error) {
      expect(error.statusCode).toBe(500)
      expect(error.error).toEqual({
        error: 'no req.body found, is app.use(bodyParser.json()) hooked up?',
      })
    }
    stopServer(server)
  })

  it('should set custom error code in request body', async () => {
    expect.assertions(1)
    const name = 'name'
    const errorMessage = 'nope'
    const errorCode = 1001
    const fn = async () => {
      const error = new Error(errorMessage)
      error.handled = true
      error.rpcError = true
      error.code = errorCode
      throw error
    }
    const method = {
      name,
      fn,
    }
    const server = createServer(rpc(method))
    try {
      let url = await listen(server)
      await generateRequest({
        url,
        name,
      })
    } catch (error) {
      expect(error.error).toEqual({
        error: errorMessage,
        code: errorCode,
        handled: true,
      })
    }
    stopServer(server)
  })

  it('should handle dependency injection for rpc methods', async () => {
    const name = 'name'
    const fn = jest.fn((req, res, { Utility }) => {
      return Promise.resolve(Utility.someFunc())
    })
    const method = {
      name,
      fn,
    }
    // method that doesn't use utils
    const fnTwo = jest.fn(() => {
      return Promise.resolve(true)
    })
    const methodTwo = {
      name: 'methodTwo',
      fn: fnTwo,
    }
    const server = createServer(
      rpc([method, methodTwo], { Utility: { someFunc: jest.fn() } }),
    )
    let url = await listen(server)
    await generateRequest({
      url,
      name,
    })
    await generateRequest({
      url,
      name: 'methodTwo',
    })
    expect(fn).toHaveBeenCalled()
    stopServer(server)
  })
})


test/index.test.js
const listen = require('test-listen')
const { createServer, stopServer, generateRequest } = require('./utils')
const { rpc, method, createError, errorMiddleware } = require('../src/')

describe('index', () => {
  it('should handle a response', async () => {
    const name = 'someMethod'
    const result = 'hello, world'
    const server = createServer(rpc(method(name, () => result)))
    let url = await listen(server)
    const body = await generateRequest({
      url,
      name,
    })
    expect(body).toEqual({ result })
    stopServer(server)
  })

  it('should handle a handled error response', async () => {
    expect.assertions(2)
    const name = 'someMethod'
    const message = 'nope'
    const server = createServer(
      rpc(
        method(name, () => {
          throw createError({
            message,
          })
        }),
      ),
    )
    let url = await listen(server)
    try {
      await generateRequest({
        url,
        name,
      })
    } catch (error) {
      expect(error.statusCode).toBe(400)
      expect(error.error).toEqual({
        error: message,
        code: 1000,
        handled: true,
      })
    }
    stopServer(server)
  })

  it('should handle a handled error response with custom code', async () => {
    expect.assertions(2)
    const name = 'someMethod'
    const message = 'nope'
    const code = 10001
    const server = createServer(
      rpc(
        method(name, () => {
          throw createError({
            message,
            code,
          })
        }),
      ),
    )
    let url = await listen(server)

    try {
      await generateRequest({
        url,
        name,
      })
    } catch (error) {
      expect(error.statusCode).toBe(400)
      expect(error.error).toEqual({
        error: message,
        code,
        handled: true,
      })
    }
    stopServer(server)
  })

  it('should handle an unhandled error response', async () => {
    expect.assertions(2)
    const name = 'someMethod'
    const message = 'nope'
    const customErrorHandler = (error, req, res, next) => {
      if (res.headersSent) {
        return next(error)
      }
      res.status(500).send({ error: error.message })
    }
    const server = createServer(
      rpc(
        method(name, () => {
          throw new Error(message)
        }),
      ),
      customErrorHandler,
    )
    let url = await listen(server)

    try {
      await generateRequest({
        url,
        name,
      })
    } catch (error) {
      expect(error.statusCode).toBe(500)
      expect(error.error).toEqual({
        error: message,
      })
    }
    stopServer(server)
  })

  it('should handle an unhandled error response with error middleware', async () => {
    expect.assertions(2)
    const name = 'someMethod'
    const message = 'nope'
    const server = createServer(
      rpc(
        method(name, () => {
          throw new Error(message)
        }),
      ),
      errorMiddleware,
    )
    let url = await listen(server)
    try {
      await generateRequest({
        url,
        name,
      })
    } catch (error) {
      expect(error.statusCode).toBe(500)
      expect(error.error).toEqual({
        error: message,
        code: 5000,
        handled: false,
      })
    }
    stopServer(server)
  })
})


src/rpc.js
/**
 * Compose our RPC Express route handler. (Typically `/rpc/:method?`).

 * If `methods` passed to this function is NOT an Array then we expect this was 
 * invoked with methods as each parameter, and so merge everything. E.g., this 
 * functions was called like so.
 * 
 *   rpc(methodOne, methodTwo, methodThree, ...);
 * 
 * Otherwise the first param should be an Array of methods and the second an
 * Object to pass as utilities / dependencies for the RPC methods. E.g.,
 *  
 *   rpc([methodOne, methodTwo, ...], { MyUtility, OtherUtil })
 * 
 * In the latter case, when RPC methods are called they will have access to the
 * utility Object via their last parameter. (See where `utils` is passed below.)
 * 
 * @returns Function  Handler for RPC requests
 */
module.exports = (methods, utils, ...rest) => {
  let allMethods = methods
  if (!Array.isArray(methods)) {
    allMethods = [methods, utils, ...rest].filter(i => i) // filter removes null / undefined
    utils = null
  }
  return (req, res, next) => {
    if (!req.body) {
      return next(
        new Error(
          'no req.body found, is app.use(bodyParser.json()) hooked up?',
        ),
      )
    }
    const { name: bodyName, args } = req.body
    // Try to get the method name from the URL first (e.g., /rpc/:method)
    // Fall back to checking header or POST body
    const name =
      req.params.method || req.headers['x-buffer-rpc-name'] || bodyName
    const matchingMethod = allMethods.find(method => method.name === name)
    if (name === 'methods') {
      res.send([
        {
          name: 'methods',
          docs: 'list all available methods',
        },
        ...allMethods.map(({ name, docs }) => ({
          name,
          docs,
        })),
      ])
    } else if (matchingMethod) {
      const parsedArgs = args ? JSON.parse(args) : []
      let promise
      try {
        const fnResult = Array.isArray(parsedArgs)
          ? matchingMethod.fn(...parsedArgs, req, res, utils)
          : matchingMethod.fn(parsedArgs, req, res, utils)
        // if async function set it to the promise
        if (fnResult.then) {
          promise = fnResult
        } else {
          // if sync function wrap in a resolved promise
          promise = Promise.resolve(fnResult)
        }
      } catch (error) {
        // sync failure, wrap in a rejected promise
        promise = Promise.reject(error)
      }
      // handle request the same for sync or async
      promise.then(result => res.send({ result })).catch(error => {
        if (error.rpcError) {
          res.status(error.statusCode || 400).send({
            error: error.message,
            code: error.code,
            handled: error.handled,
          })
        } else {
          next(error)
        }
      })
    } else {
      res.status(404).send({
        error: 'unknown method',
        code: 4040,
        handled: true,
      })
    }
  }
}
